### PR TITLE
External Table Primary Key

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes [#7154](https://github.com/apache/arrow-datafusion/issues/7154).

## Rationale for this change
With the changes in this PR, we can use use primary key information in the external tables also.

## Are these changes tested?
Yes new tests are added to show that, primary key information is used during planning. With these changes query below can run 
```sql
SELECT c, b, SUM(d)
FROM multiple_ordered_table2
GROUP BY c;
```
even if column `b` is not among the group by expressions, when it is known that column `c` is primary key. In this case we are sure that, for each unique `c` value, there will be corresponding `b` value (`b` is functionally dependent on the `c`). Hence, above query can run successfully.

### CODE REVIEW

File: datafusion/common/src/functional_dependencies.rs [hunk 1]
```diff
@@ -46,10 +47,12 @@ impl Constraints {
...
-    // This method is private.
-    // Outside callers can either create empty constraint using `Constraints::empty` API.
-    // or create constraint from table constraints using `Constraints::new_from_table_constraints` API.
-    fn new(constraints: Vec<Constraint>) -> Self {
+    /// Create a new `Constraints` from given `constraints`.
+    /// Users should use `empty` or `new_from_table_constraints`
+    /// API, for constructing `Constraints`. This API doesn't check
+    /// whether argument is valid. User is responsible with
+    /// giving correct `Vec<Constraint>`
+    pub fn new(constraints: Vec<Constraint>) -> Self {
```
Comment (mustafasrepo) [inline datafusion/common/src/functional_dependencies.rs:55]: Unfortunately, I needed to use this function during conversion from proto format.

File: datafusion/core/src/datasource/listing/table.rs [hunk 2]
```diff
@@ -590,6 +591,7 @@ pub struct ListingTable {
     definition: Option<String>,
     collected_statistics: FileStatisticsCache,
     infinite_source: bool,
+    constraints: Option<Constraints>,
```
Comment (ozankabak) [inline datafusion/core/src/datasource/listing/table.rs:594]: Let's discuss if we can represent the absence of constraints with an empty `Constraints` object.

**Review by ozankabak — APPROVED** [review]
I left one remark which we can discuss tomorrow and then we can send this upstream.

File: datafusion/core/src/datasource/listing_table_factory.rs [hunk 3]
```diff
@@ -282,6 +282,7 @@ mod tests {
             order_exprs: vec![],
             unbounded: false,
             options: HashMap::new(),
+            constraints: Constraints::empty(),
```
Comment (liukun4515) [inline datafusion/core/src/datasource/listing_table_factory.rs:285]: Do other datasource not need to add the `Constraints` field?

Comment (mustafasrepo) [reply datafusion/core/src/datasource/listing_table_factory.rs:285]: For tables that support `Constraints` we have `with_constraints` builder. Once we add support for `Constraints` for other tables. Each table will have its own `with_constraints` API to set `Constraints` for the table.

**Comment by mustafasrepo** [comment]
Merged upstream in [#7755](https://github.com/apache/arrow-datafusion/pull/7755)

**Pull request closed without merging**